[bearertokenauthextension] Load token lazily for gRPC AUTH to fix token refresh issue

ms-hujia opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
For the gRPC client AUTH part of bearertokenauthextension, there's a token refresh issue which causes all telemetry rejected by server after a period of time. The old token is always used even if the token is refreshed in filesystem. The root cause is the token value is cached during the `PerRPCCredentials()` instead of retrieving the latest. This PR aims to fix it by loading the token lazily in `GetRequestMetadata()`.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
N/A.

<!--Describe what testing was performed and which tests were added.-->
#### Testing
After I built a new docker image with this fix and tested in my environment, the token refresh issue didn't happen again after the first token expired.

<!--Describe the documentation added.-->
#### Documentation
N/A.
<!--Please delete paragraphs that you did not use before submitting.-->

Commit:
[bearertokenauthextension] Load token lazily for gRPC AUTH to fix token refresh issue

jpkrohling commented on extension/bearertokenauthextension/bearertokenauth_test.go:
```diff
@@ -18,15 +18,19 @@ import (
 )
 
 func TestPerRPCAuth(t *testing.T) {
```
I'm not confident enough on this test: it doesn't prove me that there's a bug with the current code, and thus, doesn't ensure that the bug won't happen again. Can you please add a test like the following one:

```go
func TestUpdateToken(t *testing.T) {
	// prepare
	cfg := createDefaultConfig().(*Config)
	cfg.BearerToken = "1234"

	bauth := newBearerTokenAuth(cfg, zaptest.NewLogger(t))
	assert.NotNil(t, bauth)

	perRPCAuth, err := bauth.PerRPCCredentials()
	assert.NoError(t, err)

	ctx := context.Background()
	assert.NoError(t, bauth.Start(ctx, componenttest.NewNopHost()))

	// initial token, OK
	_, err = bauth.Authenticate(ctx, map[string][]string{"authorization": {"Bearer " + "1234"}})
	assert.NoError(t, err)
	md, err := perRPCAuth.GetRequestMetadata(context.Background())
	assert.NoError(t, err)
	assert.Equal(t, map[string]string{"authorization": "Bearer " + "1234"}, md)

	// update the token
	bauth.setAuthorizationValue("5678")
	_, err = bauth.Authenticate(ctx, map[string][]string{"authorization": {"Bearer " + "5678"}})
	assert.NoError(t, err)

	md, err = perRPCAuth.GetRequestMetadata(context.Background())
	assert.NoError(t, err)
	assert.Equal(t, map[string]string{"authorization": "Bearer " + "5678"}, md)

	assert.NoError(t, bauth.Shutdown(context.Background()))
}
```

ms-hujia replied:
Of course I could add the test.

Regarding if it's an issue, `PerRPCCredentials()` would be called for only once when one exporter connects to the server endpoints so the token inside metadata is static and won't get a chance to update.

ms-hujia replied:
Added the test while I removed calls to `Authenticate()` as the issue is only about client AUTH.

Commit:
fix for PR comment

frzifus reviewed (approved):
Thanks!

frzifus commented:
@ms-hujia seems we miss a changelog entry.

You can run `make chlog-new` locally. This will create a file in `.chloggen` used to generate the changelogs for the next release. When filling the template you can re-use `36749` (number of your pr) as issue number.

Commit:
add changelog

ms-hujia commented:
> @ms-hujia seems we miss a changelog entry.
> 
> You can run `make chlog-new` locally. This will create a file in `.chloggen` used to generate the changelogs for the next release. When filling the template you can re-use `36749` (number of your pr) as issue number.

Hi @frzifus, I've added the changelog and rebased with main. Could you help to merge this PR?

ms-hujia commented:
Hi @jpkrohling and @frzifus, thanks for your review! I'm wondering when this fix could be merged and appear in the release.

ms-hujia commented:
Ping for merging...

justin-mccann commented:
@frzifus @jpkrohling I think one of you may need to trigger the checks and mark this as ready to merge. We will make use of this fix in whatever release it appears in.

CosminLazar commented:
A merge would be greatly appreciated!

songy23 commented:
CI is stuck, trying close and reopen..

songy23 commented:
please fix linter https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/13039807308/job/36378937537?pr=36749

Commit:
fix lint issue

ms-hujia commented:
> please fix linter https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/13039807308/job/36378937537?pr=36749

The lint issue is fixed in latest commit.
